nilearn/_utils/segmentation.py

```diff
@@ -0,0 +1,329 @@
+                      'found. Returning provided labels.')
+        out_labels = labels
+        return out_labels
+
+    # We take multichannel as always False since we are not strictly using
+    # for image processing as such with RGB values.
+    multichannel = False
+    if not multichannel:
+        if data.ndim < 2 or data.ndim > 3:
+            raise ValueError('For non-multichannel input, data must be of '
+                             'dimension 2 or 3.')
+        dims = data.shape  # To reshape final labeled result
+        data = np.atleast_3d(as_float_array(data))[..., np.newaxis]
+
+    # Spacing kwarg checks
+    if spacing is None:
+        spacing = np.asarray((1.,) * 3)
+    elif len(spacing) == len(dims):
+        if len(spacing) == 2:  # Need a dummy spacing for singleton 3rd dim
+            spacing = np.r_[spacing, 1.]
+        else:                  # Convert to array
+            spacing = np.asarray(spacing)
+    else:
+        raise ValueError('Input argument `spacing` incorrect, should be an '
+                         'iterable with one number per spatial dimension.')
+
+    if copy:
+        labels = np.copy(labels)
+    label_values = np.unique(labels)
+
+    # Reorder label values to have consecutive integers (no gaps)
+    if np.any(np.diff(label_values) != 1):
+        mask = labels >= 0
+        labels[mask] = np.searchsorted(np.unique(labels[mask]),
+                                       labels[mask])[0].astype(labels.dtype)
+    labels = labels.astype(np.int32)
+
+    # If the array has pruned zones, be sure that no isolated pixels
+    # exist between pruned zones (they could not be determined)
+    if np.any(labels < 0):
+        filled = ndi.binary_propagation(labels > 0, mask=labels >= 0)
+        labels[np.logical_and(np.logical_not(filled), labels == 0)] = -1
+        del filled
+    labels = np.atleast_3d(labels)
+    if np.any(labels < 0):
+        lap_sparse = _build_laplacian(data, spacing, mask=labels >= 0, beta=beta)
+    else:
+        lap_sparse = _build_laplacian(data, spacing, beta=beta)
+
+    lap_sparse, B = _buildAB(lap_sparse, labels)
+
+    # We solve the linear system
+    # lap_sparse X = B
+    # where X[i, j] is the probability that a marker of label i arrives
+    # first at pixel j by anisotropic diffusion.
+    if mode == 'cg':
+        X = _solve_cg(lap_sparse, B, tol=tol, return_full_prob=False)
+
+    # Clean up results
+    X = _clean_labels_ar(X + 1, labels).reshape(dims)
+    return X
+
+
+def _solve_cg(lap_sparse, B, tol, return_full_prob=False):
+    """
+    solves lap_sparse X_i = B_i for each phase i, using the conjugate
+    gradient method. For each pixel, the label i corresponding to the
+    maximal X_i is returned.
+    """
+    lap_sparse = lap_sparse.tocsc()
+    X = []
+    for i in range(len(B)):
+        x0 = cg(lap_sparse, -B[i].todense(), tol=tol)[0]
+        X.append(x0)
+    if not return_full_prob:
```

Comment:
We can probably remove the option return_full_prob and only keep the value that we use.